python/ray/workflow/workflow_state.py

```diff
@@ -0,0 +1,290 @@
+import asyncio
+
+from collections import deque, defaultdict
+import dataclasses
+from dataclasses import field
+import logging
+from typing import List, Dict, Optional, Set, Deque, Callable
+
+import ray
+from ray.workflow.common import (
+    Workflow,
+    StepID,
+    WorkflowRef,
+    WorkflowStepRuntimeOptions,
+)
+from ray.workflow.workflow_context import WorkflowStepContext
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class TaskExecutionMetadata:
+    submit_time: Optional[float] = None
+    finish_time: Optional[float] = None
+    output_size: Optional[int] = None
+
+    @property
+    def duration(self):
+        return self.finish_time - self.submit_time
+
+
+@dataclasses.dataclass
+class WorkflowExecutionState:
+    """The execution state of a workflow. This dataclass helps with observation
+    and debugging."""
+
+    # -------------------------------- dependencies -------------------------------- #
+
+    # The reversed dependency mapping from downstream tasks to upstream tasks.
+    in_dependency: Dict[StepID, List[StepID]] = field(default_factory=dict)
+    # The dependency mapping from upstream tasks to downstream tasks.
+    out_dependency: Dict[StepID, List[StepID]] = field(
+        default_factory=lambda: defaultdict(list)
+    )
```

Comment:
```suggestion
    # The mapping from all tasks to immediately upstream tasks.
    upstream_dependencies: Dict[StepID, List[StepID]] = field(default_factory=dict)
    # A reverse mapping of the above. The dependency mapping from tasks to immediately downstream tasks.
    downstream_dependencies: Dict[StepID, List[StepID]] = field(
        default_factory=lambda: defaultdict(list)
    )
```

Also, do these maps always hold the same tasks? When do we delete from them, if ever?

Reply:
currently we do not delete them, because in theory all of them could be used for online recovery (but yes, we could delete some of them if they are checkpointed). the whole state will be deleted when a workflow finishes